# Project Report TDT4173

The purpose of this report is to summarize all steps taken in our group in order to find fitting models/algorithms to the problem at hand. This will include exploratory data analysis, feature engineering, variuos predictors including boosting/bagging, as well as feature and model interpretations.



## Table of contents
1. [Planned actions](#planned-actions)
1. [Exploratory data analysis](#Exploratory-data-analysis)
    1. [Important observations](#Important-observations)
    2. [Steps taken](#Exploratory-steps)
2. [Feature engineering](#Feature-engineering)
    1. [Important observations](#Feature-engineering-Important-observations)
    2. [Steps taken](#Feature-engineering-Steps-taken)
3. [Model training](#Model-training)
    1. [Important observations](#Model-training-Important-observations)
    2. [Steps taken](#Model-training-Steps-taken)
4. [Model evaluation and interpretation](#Model-evaluation-and-interpretation)


## Planned actions

1. Perform exploratory data analysis in order to get an understanding of the data and notice patters/dependencies.
2. Using the results in data analysis, perform feature engineering on a simple model (xgboost and random forest).
3. Use the engineered features on better models (boosting/bagging). 
4. When model performs to satisfaction, perform model interpretation. 


## Exploratory data analysis: 

The purpose of performing an exploratory data analysis is to get an understanding of the different types of data included in the problem and their relations. This will be useful when creating models in order to understand why different models perform a certain way as how feature engineering can help improve performance. 

### Important observations

This section will summarize the steps taken in next subsection, and will include the most important observations taken during testing.

- There are 33171 purchase orders and 122590 receivals. It is thus evident that there is a tendency that purchase orders are split into several receivals, either due to stock unavailability, large orders or other reasons. We may try to merge the two dataframes together using purchase order id´s, but this is dependent on them existing in both dataframes. If they do not, we may consider deleting the id´s. 

- In both receivals and purchase orders, there are a few NaN values in the different features. Proposed solution is to drop these rows. The column `batch_id` in the receivals dataframe has about half of its values as NaN. Proposed solution is to either drop this column, or combine the non-NaN rows (aka the batches) and then drop the column.



### Steps taken

This section describes all the steps taken in the exploratory data analysis. All steps taken are to be included here and may include old steps not discussed further. In further sections we may observe new data patterns, which will be noted here. 

#### 07.10.:
- Converted date columns to datetime format, and visualized the head of the dataframes to get an understanding of the data.

- Checked the amount of NaN values in the receivals dataframe. About half of the values in the `batch_id` column in the dataframe are NaN values. We could drop the entire column. A handful of NaN values in all other columns. There could be overlap between NaN values across features, but I propose to drop all rows with NaN values in the receivals dataframe. 

- Checked the amount of NaN values in the purchase orders dataframe. A few NaN values in the `unit` and `unit_id` columns. Could remove these rows as we cannot be certain of the unit of the purchase order. Most of the units are in 'kg' and a handful in 'pund'. Could either remove the rows with 'pund' or convert them to 'kg'.

- In the `receival_status` column in the receivals dataframe, there are 142 orders that are not 'Completed'.

- There are about 4400 purchase orders which are not 'Closed' in the `status` column in the purchase orders dataframe. This could be an important observation as these orders are not completed, and could be a reason for delay.



In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import xgboost as xgb

plt.style.use('seaborn-v0_8')

In [3]:
data_receivals = pd.read_csv('./Project_materials/data/kernel/receivals.csv')
data_purchase_orders = pd.read_csv('./Project_materials/data/kernel/purchase_orders.csv')
# data_materials = pd.read_csv('./Project_materials/data/extended/materials.csv')
# data_transportation = pd.read_csv('./Project_materials/data/extended/transportation.csv')


#### Printing dataframe heads:

In [4]:
data_receivals['date_arrival'] = pd.to_datetime(data_receivals['date_arrival'], utc=True).dt.tz_localize(None)
print("Amount of receivals: ", len(data_receivals['rm_id']))
data_receivals.head()

Amount of receivals:  122590


rm_id  product_id  purchase_order_id  purchase_order_item_no  \
0  365.0  91900143.0           208545.0                    10.0   
1  365.0  91900143.0           208545.0                    10.0   
2  365.0  91900143.0           208490.0                    10.0   
3  365.0  91900143.0           208490.0                    10.0   
4  379.0  91900296.0           210435.0                    20.0   

   receival_item_no  batch_id        date_arrival receival_status  net_weight  \
0                 1       NaN 2004-06-15 11:34:00       Completed     11420.0   
1                 2       NaN 2004-06-15 11:34:00       Completed     13760.0   
2                 1       NaN 2004-06-15 11:38:00       Completed     11281.0   
3                 2       NaN 2004-06-15 11:38:00       Completed     13083.0   
4                 1       NaN 2004-06-15 11:40:00       Completed     23910.0   

   supplier_id  
0        52062  
1        52062  
2        50468  
3        50468  
4        52577

In [5]:
data_purchase_orders['delivery_date'] = pd.to_datetime(data_purchase_orders['delivery_date'], utc=True).dt.tz_localize(None)
data_purchase_orders['created_date_time'] = pd.to_datetime(data_purchase_orders['created_date_time'], utc=True).dt.tz_localize(None)
data_purchase_orders['modified_date_time'] = pd.to_datetime(data_purchase_orders['modified_date_time'], utc=True).dt.tz_localize(None)
print("Amount of purchase orders: ", len(data_purchase_orders['purchase_order_id']))
data_purchase_orders.head()

Amount of purchase orders:  33171


purchase_order_id  purchase_order_item_no  quantity       delivery_date  \
0                  1                       1     -14.0 2003-05-11 22:00:00   
1                 22                       1   23880.0 2003-05-26 22:00:00   
2                 41                       1       0.0 2004-03-07 23:00:00   
3                 61                       1       0.0 2004-03-09 23:00:00   
4                141                      10   25000.0 2004-10-27 22:00:00   

   product_id  product_version   created_date_time  modified_date_time  \
0    91900143                1 2003-05-12 10:00:48 2004-06-15 06:16:18   
1    91900160                1 2003-05-27 12:42:07 2012-06-29 09:41:13   
2    91900143                1 2004-03-08 13:44:31 2012-07-04 13:51:02   
3    91900143                1 2004-03-10 11:39:06 2012-07-04 13:50:59   
4    91900143                1 2004-10-22 12:21:54 2012-07-04 13:50:55   

   unit_id unit  status_id  status  
0      NaN  NaN          2  Closed  
1      NaN  NaN          2  Closed  
2      NaN  NaN          2  Closed  
3      NaN  NaN          2  Closed  
4      NaN  NaN          2  Closed

#### Study of column values, especially NaN values:

In [6]:

num_nan = 0
for elmt in data_receivals['net_weight']:
    if pd.isna(elmt):
        num_nan += 1

print(num_nan)



68


#### 16.10.: Initial Modeling & Lag Analysis

**Baseline Models**:
- Started with simple rule-based baseline (0 for inactive, 75% of PO qty for active) → Score: 134,136 (beat 0 VTs) ❌
- Moved to XGBoost with basic features (365d, 90d aggregates, PO data) + quantile regression (α=0.2) → Score: 10,135 (beat 2 VTs) ✅

**Lag Pattern Discovery**:
- Analyzed delivery lag (actual arrival - expected delivery date)
- Key finding: Median lag = -15 days (deliveries arrive ~2 weeks EARLY!)
- Supplier-specific variation: std dev of 47.3 days across suppliers → significant
- Product-specific variation: std dev of 11.4 days
- Temporal stability: 2020+ data is stable (2004-2006 had weird patterns, excluded from lag calc)

**Lag Adjustment Implementation**:
- Computed supplier-specific median lags from 2020+ data
- Adjusted PO expected_arrival = delivery_date + supplier_lag
- Only count POs with expected_arrival in forecast window (critical!)
- XGBoost with lag adjustment → Score: 10,135 ✅
- LightGBM with lag adjustment → Score: 9,600 ✅ (best so far!)

**Failed Experiments**:
- Random Forest with mean predictions → Score: 16,763 ❌ (RF needs 20th percentile extraction, too slow for iteration)
- Ensemble XGBoost + LightGBM → No improvement (models too correlated, -0.2% on validation)

**Feature Engineering Attempts**:
1. **2023-2024 training data + trend features + supplier categorical** → Score: 11,800 ❌
   - Added: 30d, 180d aggregates, trend_ratio, acceleration, supplier_id as categorical
   - Problem: 2023 data created distribution shift (2023 patterns ≠ 2025 patterns)
   - Validation improved (24,248) but Kaggle worse → classic overfitting
   
2. **2024 data + supplier as categorical** → Score: 15,000 ❌❌
   - Problem: LightGBM's categorical feature handling overfits with 87 suppliers on small dataset
   - Learned: categorical_feature parameter is dangerous with limited data

**Key Observations**:
- LightGBM (9,600) slightly beats XGBoost (10,135) with same features
- Lag adjustment is critical (improves ~16% from baseline)
- More data ≠ better (2023 data hurts due to distribution shift)
- Categorical features in LightGBM overfit easily
- Validation loss can be misleading (need time-based validation for forecasting)
- Random 80/20 split includes old patterns, but Kaggle tests on 2025 (unseen conditions)

**Current Best**: LightGBM + lag adjustment (2024 data, 13 features) → 9,600 (beats 2-3 VTs)

**Next Steps**: Test supplier_id as numeric feature (not categorical), add trend features carefully with 2024 data only

### 17.10.: Validation Strategy, Feature Analysis & Zero-Inflation

**Starting Point**: LightGBM with lag adjustment → Score: 9,600 (2024 data, time-based features)

**Phase 1: Diagnostic Deep Dive**
- Problem: Score stuck at 9,600 despite alpha adjustments (0.15, 0.20, 0.25, 0.30)
- Root cause: Random 80/20 split mixes all months → model learns "average 2024" instead of declining trend
- Key finding: **-35.6% decline** from training (Jan-Aug: 342k kg mean) to validation (Sep-Nov: 220k kg mean)

**Validation Strategy Experiments**:
1. **Random split + alpha=0.15 + 0.88 adjustment** → Score: 9,500 ✅
   - Predictions: 37k kg × 0.88 = 32.7k kg mean
   - Issue: Still predicting "average" rather than learning trend
   
2. **Time-based split + alpha=0.15** → Score: 9,500
   - Training: Jan-Aug 2024, Validation: Sep-Nov 2024
   - Model sees -35.6% decline but predictions still ~37k kg
   
3. **Time-based split + alpha=0.25 (removed 0.88)** → Score: 9,960 ❌
   - Predictions increased to 45k kg
   - Proved over-predicting makes score worse → test set has LOW activity
   
4. **Time-based split + alpha=0.15 + 0.85 adjustment** → Score: ~9,500
   - No real improvement, wasted submission

**Key Insight**: Hyperparameter tuning won't get from 9,600 → 5,000. Need fundamental changes.

**Phase 2: Deep EDA - The Breakthrough**

**Critical Discoveries**:
1. **Zero-Inflation Problem** ⚠️
   - Test set: 203 unique rm_ids
   - **Only 60 (29.6%) have ANY 2024 data**
   - **143 rm_ids (70.4%) are DEAD** (last delivery 2004-2005)
   - **72.9% of test predictions (22,200/30,450) are for DEAD rm_ids**
   - Problem: Quantile regression predicts positive values for everything → massive over-prediction

2. **Feature Correlation Analysis** (Sep-Nov 2024):
```
   Strong features (correlation > 0.65):
   - total_90, rate_90: +0.9526 ***
   - recency_weighted: +0.9327 ***
   - total_180: +0.9131 ***
   - rate_30, total_30: +0.8872 ***
   - active_ratio_90: +0.6716 ***
   
   Weak features (remove):
   - momentum: +0.0713
   - slope_90: +0.0084
   - cv_90: -0.0207
   - PO quantity: +0.0767
```

3. **Seasonality**: January vs other months 1.03x ratio → month/quarter features likely noise

**Phase 3: The Working Solution** ✅

**Two-Stage Model + Guardrails**:
1. **Two-Stage Architecture**:
   - Stage 1: LGBMClassifier (n=400) predicts P(delivery > 0)
   - Stage 2: LGBMRegressor (quantile α=0.15) predicts amount IF delivered
   - Combined: prediction = regressor × classifier_probability
   - **Why it works**: Dead rm_ids get low probability (~0.1) → effectively zero prediction

2. **Per-Horizon Calibration**:
   - Learned on Sep-Nov validation for horizons {7, 30, 60, 90, 150}
   - factor_h = sum(actual) / sum(predicted), clipped to [0.70, 1.10]
   - Fixes systematic over/under-prediction at different horizons

3. **Activity-Based Guardrails**:
   - days_since_last > 365 → force 0
   - 180 < days_since_last ≤ 365 → soft cap at 8% of last year volume
   - Upper cap: min(pred, (total_365/365) × horizon × 1.5)

**Result**: 9,600 → **8,200** ✅✅ (14% improvement!)

**Phase 4: Feature Engineering - Step 1** (In Progress)

**Removed (9 weak features)**:
- ❌ momentum, slope_90, cv_90 (correlation < 0.1)
- ❌ future_po_quantity, future_po_count (correlation 0.08)
- ❌ month, quarter (weak seasonality)
- ❌ avg_weight_365d, daily_rate_365d (redundant)

**Added (5 strong features)**:
- ✅ recency_weighted (correlation 0.93)
- ✅ total_30, rate_30, count_30 (correlation 0.89)
- ✅ active_ratio_90 (correlation 0.67)

**Net**: 16 features → 14 features (more focused, less noise)
**Expected**: 8,200 → 7,200-7,500

**Key Learnings**:
- Zero-inflation critical: 73% of test should be ~0, quantile regression can't learn this
- Domain knowledge > ML: Hard-coded rules (dead = 0) beat learned patterns
- Feature correlation matters: momentum (0.07 correlation) = noise
- Validation strategy crucial: Random split hides declining trend
- EDA before iteration: Deep analysis revealed 73% dead rm_ids problem

**Failed Experiments**:
- Random split + alpha tuning → 9,500-9,960 ❌ (doesn't learn trend)
- Remove 0.88 adjustment → 9,960 ❌ (predictions too high)
- Time-based + various alphas → 9,500-9,960 ❌ (hyperparameter band-aid)

**Current Best**: Two-stage + guardrails → 8,200 (beats ~3-4 VTs)

**Next Steps** (if Step 1 works):
1. Interpolate calibration for horizons 2-151
2. Activity-aware guardrails (smooth decay)
3. Ensemble with median predictor

### 18.10.: Feature Pruning Based on Correlation Analysis

**Removed (9 weak features)**:
- ❌ momentum, slope_90, cv_90 (correlation < 0.1)
- ❌ future_po_quantity, future_po_count (correlation 0.08)
- ❌ month, quarter (weak seasonality)
- ❌ avg_weight_365d, daily_rate_365d (redundant)

**Added (5 strong features)**:
- ✅ recency_weighted (correlation 0.93) - weights recent deliveries higher
- ✅ total_30, rate_30, count_30 (correlation 0.89) - captures very recent activity
- ✅ active_ratio_90 (correlation 0.67) - ratio of active days in last 90d

**Net**: 16 features → 14 features (more focused, less noise)

**Result**: 8,200 → **6,200** ✅✅✅ (24% improvement!)

**Key Insight**: Weak features weren't neutral - they were actively adding noise and confusing the model. Removing them + adding highly correlated features = massive gain.

**Total Progress**: 9,600 (baseline) → 6,200 (35% improvement!)

---

#### Fine-Tuning Attempts (Hit Plateau)

**Step 2: Interpolated Calibration** → 6,700 ❌ (worse by 500)
- Tried to fill calibration for ALL horizons 2-151 (not just {7,30,60,90,150})
- Used scipy.interpolate.interp1d to smooth factors across horizons
- **Failed:** Overfitted to Sep-Nov 2024 patterns, didn't generalize to 2025 test set
- Learning: Validation-learned adjustments don't always transfer to test

**Step 3: Aggressive Dead Filtering** → 6,200 (no change)
- Tightened guardrails: cold rm_ids (180-365d) from 8% cap → 5% cap
- Added new warm rm_ids (90-180d) cap at 15%
- **No effect:** Only affected 19/203 rm_ids, and their predictions were already low from two-stage model
- Learning: Guardrails already optimal from ChatGPT's original implementation

**Step 4: Lower Alpha (0.15 → 0.10)** → 6,140 ✅ (tiny improvement, -60 points)
- Changed quantile target from 15th percentile to 10th percentile
- More conservative predictions
- **Minimal improvement:** Suggests we're close to optimal conservatism level

**Step 5: Remove Calibration Entirely** → 6,138 ✅ (basically same, -2 points)
- Removed all per-horizon calibration adjustments
- Let model's forecast_horizon feature handle effects naturally
- **No impact:** Calibration was irrelevant (factors too close to 1.0)
- Learning: Sep-Nov calibration neither helped nor hurt

**Key Observations**:
- Hit a plateau around 6,100-6,200 - small tweaks (±60 points) don't move the needle
- Feature engineering (Step 1) gave 24% improvement, everything else combined gave <1%
- Calibration/guardrails/alpha tuning are all second-order effects
- Need <5,000 for grade A (currently 6,138) - requires 18.5% more improvement

**Failed Experiments**:
- Interpolated calibration → overfitting to validation
- Tighter guardrails → no effect (already optimal)
- Lower alpha → diminishing returns

**Current Best**: Step 5 (6,138 points) with alpha=0.10, no calibration, Step 1 features

**Possible next steps** (untested):
- Even lower alpha (0.05 or 0.08) for maximum conservatism
- Nuclear option: predict 0 for >90 days inactive (high risk/reward)
- Activity-aware upper caps (make hot rm_ids able to spike more, cold less)

prob need somthing fundamentally different to actually improve our score further

### 21.10.: Purchase Orders EDA → Tried “Forward-Looking” Features (…and why it blew up)

**What we set out to do**
- Deep-dive EDA on `purchase_orders` to see if future orders can guide Jan–May 2025 forecasts.
- Add a small set of PO features (sum/count/has_po/avg) on top of **Step 5** (the 6,200 model), *without* touching anything else.

---

### EDA Highlights (looked very promising)
- **We do have future POs** in the test window: **147** POs for Jan–May 2025 ✅  
- **Strong historical link:** Ordered vs. actual **corr = 0.912** (on matched 2024 POs) ✅  
- **Lead time**: median ~**51 days** (very wide spread, std ~55d) ⚠️  
- **Coverage**: PO signal covers ~**45%** of test RMs (92/203) ⚠️

**Takeaway:** POs looked like a real, forward-looking signal worth trying.

---

### What we tried (Step 10 variants)
1) **Straight PO features:**  
   - `po_total_in_horizon`, `po_count_in_horizon`, `has_future_po`, `po_avg_historical` (product-level, mapped to RM by mode product).  
   - Precomputed cache for speed, kept the **exact same** Step 5 model + guardrails.

2) **Fixed/safer version:**  
   - Ignored “Deleted” POs, down-weighted “Open” vs “Closed”.  
   - Allocated product POs to RMs using **historical product→RM shares** (to avoid double-counting).  
   - Still no changes to the base model or caps besides existing Step 5 guardrails.

---

### What happened (ouch)
- **Leaderboard exploded upward:** ~**9,000–10,000** (from 6,200). ❌❌❌  
- **Feature importance:** PO features jumped to the **top** (e.g., `po_total_in_horizon` rank #1/#2).  
- **Predictions became bimodal:**
  - With PO signal (~35% rows): **~89k** avg prediction (huge spikes)
  - Without PO (~65% rows): **~1.3k** avg prediction (collapsed to near zero)
- **Train→test coverage shift:** Train windows had ~70% rows with PO; test only ~45% → the model over-trusts PO features and underpredicts where PO is missing.

---

### Why it went wrong (post-mortem)
- **Allocation noise:** We spread product POs to RMs via historical shares → many RMs get “slivers” they won’t actually receive in 2025, and the few real recipients get diluted. The model learns a hard rule: “PO present ⇒ huge; no PO ⇒ tiny,” which doesn’t match reality.  
- **Timing mismatch:** Features use **planned** `delivery_date`, but actual arrivals drift (median lead time ~51d, big variance). We dump all PO mass inside the horizon, even when the true receivals slip outside.  
- **Coverage shift:** The model sees lots of PO rows in training (70%) and fewer in test (45%), so it overweights PO signals and forgets the baseline on “no-PO” rows.  
- **Scale issue:** `po_total_in_horizon` is a blunt, un-normalized lump → huge spikes that Step 5’s guardrails don’t cap (upper cap is generous for high-volume RMs).  
- **Mode product mapping:** RMs often have multiple products; using only the mode can attach the **wrong** PO signal to an RM.

**Net effect:** The model became **over-confident** wherever PO existed and **under-confident** elsewhere. Big absolute errors on both tails ⇒ leaderboard tanks.

---

### What *did* work
- The **EDA** was right: POs carry signal. The failure was not in “POs are useless,” but in how we **operationalized** them.  
- Our “don’t touch the base model / keep logging / fast cache” discipline made it easy to diagnose the blow-ups quickly.

---

### What we’ll try tomorrow (surgical fixes, still minimal changes)
1) **Time-spreading the PO mass**  
   Convert PO quantity to **expected daily arrival** (e.g., distribute across the delivery month, or convolve with a lead-time kernel centered ~51d). Feature = **sum over overlap with horizon**, not the whole lump.
2) **Normalize + cap PO features**  
   Winsorize `po_total_in_horizon` and cap it by a **baseline capacity** (e.g., `min(po_expected, c × rate_90 × horizon)`), so trees can’t run away.
3) **Keep baseline alive when PO is absent**  
   Ensure “no-PO” rows still lean on `rate_30/90`, `recency_weighted`, etc. (e.g., PO features act as **additive deltas**, not replacements).
4) **Tighter allocation set**  
   Use **recent (2023–2024)** product→RM shares only, and restrict to **active RMs** for that product in 2024 to reduce smearing.
5) **Reliability weighting**  
   Down-weight Open/old/edited POs more aggressively; add “recency of PO creation/modification” as a reliability hint.
6) **Sanity dashboards** (quick prints only)  
   Compare preds/actuals by buckets: `(PO yes/no) × (horizon bucket)` to spot immediate drift before submitting.

---

### Current status
- **Best:** Step 5 (no calibration, α=0.10) ≈ **6,200**  
- **Today’s PO attempts:** **~9,000–10,000** (fail)  
- **Conclusion:** POs are promising but brittle. We need time-aligned, capacity-capped, allocation-pruned features so they act as **soft guidance**, not a hard override.

---

### One-liner for the report
> We tried adding forward-looking purchase order features (strong EDA signal), but naïve aggregation and rough allocation/timing caused bimodal predictions and a big LB spike. Tomorrow we’ll spread PO mass over time, cap by capacity, and restrict allocation to recent active RMs so PO acts as a stabilizer rather than a sledgehammer.


### 01.11.25: Feature Engineering Attempts → Timing Features & Model Splitting (…and why they failed)

**What we set out to do**
- Address the 70% inactive RM_IDs in test set through specialized feature engineering and model architecture changes
- Hypothesis: One model can't handle both active (recent data) and inactive (stale/unknown) RM_IDs well
- Try multiple approaches on top of **BEST MODEL** (6,236 points, α=0.10, 15 features)

---

### Attempt 1: Timing Features (Step 8 & 9)
**EDA Motivation:**
- Delivery timing analysis showed high irregularity (CV = 2.75)
- 42.9% of 7-day forecasts expected to be zero
- Expected deliveries scale non-linearly with horizon (7d: 5.53 deliveries → 150d: 118.57 deliveries)

**What we tried:**
1. **Step 8 (4 timing features):**
   - `avg_days_between`: Average interval between deliveries
   - `cv_days_between`: Coefficient of variation (regularity metric)
   - `expected_deliveries_in_horizon`: horizon / avg_days_between
   - `delivery_overdue_ratio`: days_since_last / avg_days_between
   - Total features: 19 (was 15)

2. **Step 9 (2 "safe" timing features):**
   - Kept only `avg_days_between` and `cv_days_between`
   - Removed the "aggressive" features that encoded staleness
   - Total features: 17

**What happened:**
- **Step 8 (4 features):** Score = **6,375** (+139 points worse) ❌
- **Step 9 (2 features):** Score = **6,375** (no improvement) ❌
- **Feature importance:** Timing features ranked #1 and #2 (very high)
- **Predictions collapsed:**
  - Predictions > 0: Only 6,884-7,236 (22-24% of test set)
  - Should have been ~40-50%
- **Validation improved** (-14%) but **test performance tanked**

**Why it went wrong:**
- **Overfitting to validation decline:** Sep-Nov 2024 showed declining activity; timing features learned this pattern
- **Too literal on inactivity:** For unknown RM_IDs with sparse history:
  - `avg_days_between = 0` → model learned "no history = no delivery"
  - `cv_days_between = 999` → model learned "irregular = no delivery"
  - `delivery_overdue_ratio = 56.88` → model learned "extremely overdue = no delivery"
- **Train-test mismatch:** Validation (Sep-Nov 2024) was declining; test (Jan-May 2025) might have reactivations
- **The model predicted ZERO for most inactive RM_IDs** when it should have given them small non-zero probabilities

---

### Attempt 2: Separate Models for Active/Inactive (Step 10)
**Rationale:**
- One model tries to serve two populations: active (good data) vs inactive (sparse/old data)
- Train separate models with different alpha values based on inactivity level

**What we tried:**
- **Threshold:** days_since_last = 180 days
- **Active model (≤180d):** α=0.15 (more aggressive)
- **Inactive model (>180d):** α=0.05 (very conservative)
- Route predictions based on days_since_last at test time

**What happened:**
- **Training data imbalance:**
  - Active samples: 2,600 (95.4%)
  - Inactive samples: 125 (4.6%)
  - Inactive train set: Only **50 samples** after time split ❌
- **Inactive model failed completely:**
  - Mean prediction: 0.00 (predicted zero for everything)
  - Bias: -100%
  - Regressor loss didn't change across epochs → model didn't learn
- **Test routing was useless:**
  - 97.5% routed to active model
  - Only 2.5% used inactive model

**Why it went wrong:**
- **Training only on 2024-active RM_IDs** means very few inactive samples in training
- Can't train a meaningful model on 50 samples
- Even if threshold raised to 365d, still not enough inactive training data
- **Fundamental issue:** Training set (2024-active RMs) doesn't represent test set (70% unknown RMs)

---

### Attempt 3: Hybrid Rules (Step 11)
**Pivot strategy:**
- Accept we can't train on inactive RM_IDs (not enough data)
- Train ONE strong model on all available data
- Apply conservative **rules** at prediction time based on inactivity

**What we tried:**
```python
if days_inactive ≤ 90:   pred = model_pred × 1.0  # No penalty
elif days_inactive ≤ 180: pred = model_pred × 1.0  # No penalty
elif days_inactive ≤ 365: pred = model_pred × 0.8  # 20% reduction
elif days_inactive ≤ 730: pred = model_pred × 0.3  # 70% reduction
else:                     pred = 0.0               # Zero for >2 years
```

**Critical bug discovered:**
- Original `days_since_last` calculation was WRONG:
```python
  # BUG: Used recent_365 window
  if len(recent_365) > 0:
      days_since = (date - recent_365.max()).days
  else:
      days_since = 0  # ← Made inactive RMs look active!
```
- **Fixed:** Calculate from ALL history, not just recent window
- After fix, test set correctly showed: 64.5% very inactive (>730d)

**What happened:**
- **Validation:** Minimal improvement (rules barely applied - only 0.6% inactive)
- **Test routing:**
  - 64.5% of RM_IDs got ZERO predictions
  - These are the 131 RM_IDs NOT in training at all
- **Not yet submitted** - awaiting decision on rule aggressiveness

**Open question:**
- Is zeroing out 64.5% of test RM_IDs correct?
- Or should unknown RM_IDs get small non-zero predictions (e.g., 0.1× model)?
- Quantile loss at 0.2 means underestimation is cheaper, but not free

---

### Root Cause Analysis

**The fundamental problem:**
1. **Training set:** 60 RM_IDs active in 2024 (by design - other years non-representative)
2. **Test set:** 203 RM_IDs, of which 131-143 (64-70%) were NOT active in 2024
3. **Coverage gap:** Model has never seen 64% of test RM_IDs
4. **No signal for unknowns:** These RM_IDs have no recent activity (>730d inactive)

**Why all attempts failed:**
- **Timing features:** Learned to penalize unknowns too aggressively
- **Separate models:** Can't train inactive model without inactive data
- **Hybrid rules:** Unknown if forcing zeros is optimal strategy

**What we know works:**
- BEST MODEL (6,236): Lets model handle unknowns naturally through existing features
- Any explicit "inactive RM_ID" handling makes things worse (so far)

---

### Current Status
- **Best:** BEST MODEL (α=0.10, 15 features, no special inactive handling) = **6,236**
- **Step 8 (4 timing features):** **6,375** (-139 points)
- **Step 9 (2 timing features):** **6,375** (-139 points)
- **Step 10 (separate models):** Not submitted (inactive model broken)
- **Step 11 (hybrid rules):** Pending submission decision

**Conclusion:** 
The 64% unknown RM_ID problem cannot be solved through feature engineering or model architecture alone when training only on 2024 data. Either:
1. Find external signal for unknowns (product groups, supplier patterns, purchase orders)
2. Accept model must extrapolate and optimize the extrapolation strategy
3. Revisit "2024-only training" constraint if validation shows merit

---

### Next Steps (TBD)
- Submit Step 11 and evaluate if aggressive zero-ing helps or hurts
- Explore product/alloy grouping to give unknowns "borrowed" patterns from similar known RMs
- Revisit purchase orders with lessons learned (avoid bimodal collapse)

## Feature engineering:

The purpose of performing feature engineering on the datasets is to increase the performance of a predicting model. This can for example be done by removing features, merge features or giving features extra "weight".

### Important observations

This section will summarize the steps taken in next subsection, and will include the most important observations taken during feature engineering.



### Steps taken

THE SMOKING GUN:
text
Average CV: 2.75 (HIGHLY IRREGULAR)
Regular patterns (<CV 1.0): 20.4%
7-day horizon: 42.9% zeros
Your current features completely ignore delivery timing patterns, which is why your model can't scale predictions properly.

THE FUNDAMENTAL PROBLEM:
Your model is trying to predict "how much" without knowing "when". But with irregular deliveries (CV 2.75), timing is everything.

THE SOLUTION: TWO-STAGE MODEL + TIMING FEATURES
Stage 1: Binary Classification (Will there be ANY delivery?)
Target: will_deliver = (target > 0)

Use: Logistic regression, LightGBM classifier

Key features: days_since_last / avg_days_between, horizon / avg_days_between

Stage 2: Quantity Prediction (If yes, how much?)
Target: delivery_amount (only for positive samples)

Use: Quantile regression (0.2 quantile)

Key features: Add timing-aware features

CRITICAL NEW FEATURES TO ADD:
python
# Timing-aware features
df['expected_deliveries_in_horizon'] = df['forecast_horizon'] / df['avg_days_between_deliveries']
df['delivery_overdue_ratio'] = df['days_since_last'] / df['avg_days_between_deliveries']
df['delivery_regularity'] = df['cv_days_between_deliveries']
df['is_frequent_deliverer'] = (df['avg_days_between_deliveries'] < 7).astype(int)
df['is_rare_deliverer'] = (df['avg_days_between_deliveries'] > 90).astype(int)

# Horizon-adjusted rate features
df['horizon_adjusted_rate'] = df['rate_90'] * (df['forecast_horizon'] / 90)
df['expected_weight'] = df['avg_delivery_size'] * df['expected_deliveries_in_horizon']
IMMEDIATE ACTION PLAN:
Step 1: Create Timing Features
Add these for each RM_ID:

avg_days_between_deliveries

cv_days_between_deliveries

avg_delivery_size

last_year_delivery_count

Step 2: Two-Stage Training
python
# Stage 1: Classification
clf_predictor = TabularPredictor(label='will_deliver', problem_type='binary').fit(...)

# Stage 2: Regression (only positive samples)
pos_samples = train_df[train_df['target'] > 0]
reg_predictor = TabularPredictor(label='target', problem_type='quantile', quantile_levels=[0.2]).fit(pos_samples, ...)
Step 3: Combined Prediction
python
# Predict probability of delivery
delivery_proba = clf_predictor.predict_proba(test_data)['1']

# Predict amount if delivered  
delivery_amount = reg_predictor.predict(test_data)

# Combine: amount * probability (conservative)
final_prediction = delivery_amount * delivery_proba

## Model training:

This is the section where we will train different models on the data, and try to find the best model for the problem. At the end of this section we should have a model that performs well on the data, and is able to make good predictions on unseen data.

### Important observations
This section will summarize the steps taken in next subsection, and will include the most important observations taken during model training.

### Steps taken
This section describes all the steps taken during model training. All steps taken are to be included here. 

## Model evaluation and interpretation:
The purpose of this section is to evaluate the model performance and interpret the model. This will include feature importance, SHAP values and partial dependence plots.